cpp/include/cudf/detail/scatter.cuh

```diff
@@ -37,13 +57,13 @@ auto scatter_to_gather(MapIterator scatter_map_begin,
 {
   using MapValueType = typename thrust::iterator_traits<MapIterator>::value_type;
 
-  static_assert(std::is_signed<MapValueType>::value,
-                "Need different invalid index if unsigned index types are added");
-  auto const invalid_index = static_cast<MapValueType>(-1);
+  // The gather_map is initialized with gather_rows value to identify pass-through entries
+  // when calling the gather_bitmask() which applies a pass-through whenever it finds a
+  // value outside the range of the target column. And the gather_rows value here will
+  // always be outside the valid range.
```

Comment:
This is true as long as they don't use a type with fewer than 31 bits as their map value type! If they use, say UINT16 then they could have `gather_rows = 65536` and when they write this sentinel value to their map it will be `0`.  This would be a user error, but it would be hard to diagnose.

You could `CUDF_EXPECTS(gather_rows < (1 << sizeof(MapValueType)), "Number of rows in gather map too large for gather map type")` or something.

Reply:
So I've realized we have a unique opportunity here. Since the map values are indices and therefore must be in the range of `size_type`, we can safely convert the scatter map type to `size_type` instead when creating the gather map. 
```
auto gather_map = rmm::device_vector<size_type>(gather_rows, gather_rows);
```
The maximum number of rows is therefore `std::numeric_limits<size_type>::max()` and this value is always outside the indexable range. So it is safe to use the `gather_rows` as a pass-through in the `size_type` gather map.